### ISSUE TITLE
Dask client connects to PBS workers, then rapidly loses them

### ISSUE THREAD

**Issue by vprzybylo** [dask/dask-jobqueue#151]
Similar to #30, I have a bash script created to run a python job on Cheyenne using a scheduler file and a 'start-workers.py' script but after the workers come online and the main script is run, the workers die off rapidly. 

I have tried using qsub to start an interactive job for jupyter lab in prior attempts (see pangeo issue #352) but if the connection is dropped, the job disappears from the queue and I am unable to ssh back in to receive the output data.  For this reason, I believe it is better to submit a job via a scheduler file.  

Here is an outline of the steps I have taken to submit the job via bash script.  Note I have changed the number of threads, processes, memory, etc with no luck in keeping the workers online.  I simply copied the arguments from the jupyter lab cell that was working fine.

This is the main bash script that calls the other scheduler/worker scripts:
#!/bin/bash
set -e

echo "Launching dask scheduler"
s=$(qsub launch-dask-scheduler.sh)
sjob=${s%.*}
echo ${sjob}

block until the scheduler starts
while true; do
    status=`qstat ${sjob} | tail -n 1`
    echo ${status}
    if [[ ${status} =~ " R " ]]; then
        break
    fi
    sleep 1
done

commented out until done
workers=${1:-10}
echo "Launching dask workers (${workers})"
for i in $(seq 1 ${workers}); do
    echo 'launching dask worker'
    wjob=$(qsub launch-dask-worker.sh)
done

source activate pangeo
python start_workers.py

status=`qstat ${wjob} | tail -n 1`
block until the workers start
while true; do
    status=`qstat ${wjob} | tail -n 1`
   
    echo ${status}
    if [[ ${status} =~ " R " ]]; then
        break
    fi
    sleep 1
done
echo 'running IPAS script'
source activate pangeo
python ./Agg_Main_phiarray.py

################################

launch-dask-scheduler.sh script:
#!/bin/bash
#PBS -N dask-scheduler
#PBS -q economy
#PBS -A UALB0017
#PBS -l select=1:ncpus=36:mpiprocs=2:ompthreads=4:mem=109G
#PBS -l walltime=05:00:00
#PBS -j oe

Setup Environment
module purge
source activate pangeo

SCHEDULER=/glade/scratch/$USER/scheduler.json
rm -f $SCHEDULER
mpirun --np 9 dask-mpi --nthreads 4 \
    --memory-limit 22e9 \
    --interface ib0 \
    --local-directory /glade/scratch/$USER \
    --scheduler-file=$SCHEDULER

#######################################
start_workers.py script:
from dask_jobqueue import PBSCluster
from dask.distributed import Client
cluster = PBSCluster(processes=18,
                     cores=8,
                     memory="6GB",
                     project='UALB0017',
                     queue='economy',
                     resource_spec='select=1:ncpus=36:ompthreads=4:mem=109G',
                     walltime='05:00:00')

c = cluster.start_workers(10)
client = Client(cluster)

#######################################
launch-dask-worker.sh (not using this currently, instead using the python start_workers.py script)
#!/bin/bash
#PBS -N dask-workers
#PBS -q economy
#PBS -A UALB0017
#PBS -l select=1:ncpus=36:mpiprocs=2:ompthreads=4
#PBS -l walltime=05:00:00
#PBS -j oe

Setup Environment
module purge
source activate pangeo

SCHEDULER=/glade/scratch/$USER/scheduler.json
mpirun --np 9 dask-mpi --nthreads 4 \
    --memory-limit 22e9 \
    --interface ib0 \
    --no-scheduler --local-directory /glade/scratch/$USER \
    --scheduler-file=$SCHEDULER

########################################

Here is a log file for a dask worker:
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.148.9.94:46539'
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.148.9.94:59753'
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.148.9.94:58225'
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.148.9.94:55667'
distributed.diskutils - INFO - Found stale lock file and directory '/gpfs/u/home/przybylo/IPAS_array/dask-worker-space/wor
ker-c0kfm5ja', purging
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:52521
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:52521
distributed.worker - INFO -              nanny at:          10.148.9.94:32798
distributed.worker - INFO -              bokeh at:           10.148.9.94:8789
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:43057
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:43057
distributed.worker - INFO -              nanny at:          10.148.9.94:43720
distributed.worker - INFO -              bokeh at:          10.148.9.94:36811
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:58792
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-qg0ew52i
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:58792
distributed.worker - INFO -              nanny at:          10.148.9.94:59953
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:41574
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -              bokeh at:          10.148.9.94:50134
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:41574
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -              nanny at:          10.148.9.94:37153
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -              bokeh at:          10.148.9.94:33988
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-pa63sigd
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:41721
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:41721
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-hdjn52na
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -              nanny at:          10.148.9.94:50370
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -              bokeh at:          10.148.9.94:56406
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-viv52842
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
distributed.diskutils - INFO - Found stale lock file and directory '/gpfs/u/home/przybylo/IPAS_array/dask-worker-space/wor
ker-vo2pubtu', purging
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-j8_v1p2h
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:41047
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:41047
distributed.worker - INFO -              nanny at:          10.148.9.94:46539
distributed.worker - INFO -              bokeh at:          10.148.9.94:53889
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-y27xg4wq
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:49268
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:49268
distributed.worker - INFO -              nanny at:          10.148.9.94:33282
distributed.worker - INFO -              bokeh at:          10.148.9.94:38760
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-6rath3fb
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:36633
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:36633
distributed.worker - INFO -              nanny at:          10.148.9.94:39102
distributed.worker - INFO -              bokeh at:          10.148.9.94:45682
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-s1fr0lv6
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:42357
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:54327
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:42357
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:54327
distributed.worker - INFO -              nanny at:          10.148.9.94:58225
distributed.worker - INFO -              nanny at:          10.148.9.94:58002
distributed.worker - INFO -              bokeh at:          10.148.9.94:45961
distributed.worker - INFO -              bokeh at:          10.148.9.94:52568
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-imugcrhu
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-31s8103f
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:47381
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:47381
distributed.worker - INFO -              nanny at:          10.148.9.94:59753
distributed.worker - INFO -              bokeh at:          10.148.9.94:53258
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-i7fldc3s
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:49428
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:46224
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:49428
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:46224
distributed.worker - INFO -              nanny at:          10.148.9.94:37400
distributed.worker - INFO -              nanny at:          10.148.9.94:55009
distributed.worker - INFO -              bokeh at:          10.148.9.94:36557
distributed.worker - INFO -              bokeh at:          10.148.9.94:45115
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:39940
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:39940
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:58470
distributed.worker - INFO -              nanny at:          10.148.9.94:49486
distributed.worker - INFO -              bokeh at:          10.148.9.94:46379
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:58470
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO -              nanny at:          10.148.9.94:46733
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -              bokeh at:          10.148.9.94:55313
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:38657
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-mbj0bzhp
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:    tcp://10.148.9.94:48927
distributed.worker - INFO -          Listening to:    tcp://10.148.9.94:48927
distributed.worker - INFO -              nanny at:          10.148.9.94:55667
distributed.worker - INFO -              bokeh at:          10.148.9.94:36750
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          4
distributed.worker - INFO -                Memory:                  333.33 MB
distributed.worker - INFO -       Local Directory: /gpfs/u/home/przybylo/IPAS_array/dask-worker-space/worker-ldadnvth
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.7:44783
distributed.nanny - INFO - Closing Nanny at 'tcp://10.148.9.94:46539'
distributed.worker - INFO - Stopping worker at tcp://10.148.9.94:41047
distributed.nanny - INFO - Closing Nanny at 'tcp://10.148.9.94:49486'`

###########################################
And a log file for the scheduler:
`distributed.nanny - INFO -         Start Nanny at: 'tcp://10.148.9.31:41209'
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.148.9.31:53305'
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.148.9.31:50455'
[cli_6]: write_line error; fd=29 buf=:cmd=init pmi_version=1 pmi_subversion=1
:
system msg for write_line failure : Bad file descriptor
[cli_6]: Unable to write to PMI_fd
[cli_6]: write_line error; fd=29 buf=:cmd=get_appnum
:
system msg for write_line failure : Bad file descriptor
Fatal error in PMPI_Init_thread: Other MPI error, error stack:
MPIR_Init_thread(474): 
MPID_Init(152).......: channel initialization failed
MPID_Init(426).......: PMI_Get_appnum returned -1
[cli_6]: write_line error; fd=29 buf=:cmd=abort exitcode=1094159
:
system msg for write_line failure : Bad file descriptor
distributed.nanny - WARNING - Worker process 21557 was killed by unknown signal
[cli_1]: Error reading initack on 8
Error on readline:: Bad file descriptor
[cli_1]: readline failed
[cli_1]: readline failed
[cli_1]: readline failed
[cli_1]: readline failed
Fatal error in PMPI_Init_thread: Other MPI error, error stack:
MPIR_Init_thread(474): 
MPID_Init(152).......: channel initialization failed
MPID_Init(426).......: PMI_Get_appnum returned -1
[cli_1]: readline failed
distributed.nanny - WARNING - Worker process 21559 exited with status 8388357179234872675
[cli_4]: write_line error; fd=21 buf=:cmd=init pmi_version=1 pmi_subversion=1
:
system msg for write_line failure : Bad file descriptor
[cli_4]: Unable to write to PMI_fd
[cli_4]: write_line error; fd=21 buf=:cmd=get_appnum
:
system msg for write_line failure : Bad file descriptor
Fatal error in PMPI_Init_thread: Other MPI error, error stack:
MPIR_Init_thread(474): 
MPID_Init(152).......: channel initialization failed
MPID_Init(426).......: PMI_Get_appnum returned -1
[cli_4]: write_line error; fd=21 buf=:cmd=abort exitcode=1094159
:
system msg for write_line failure : Bad file descriptor
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker process 21561 was killed by unknown signal
[cli_5]: write_line error; fd=25 buf=:cmd=init pmi_version=1 pmi_subversion=1
:
system msg for write_line failure : Bad file descriptor
[cli_5]: Unable to write to PMI_fd
[cli_5]: write_line error; fd=25 buf=:cmd=get_appnum
:
system msg for write_line failure : Bad file descriptor
Fatal error in PMPI_Init_thread: Other MPI error, error stack:
MPIR_Init_thread(474): 
MPID_Init(152).......: channel initialization failed
MPID_Init(426).......: PMI_Get_appnum returned -1
[cli_5]: write_line error; fd=25 buf=:cmd=abort exitcode=1094159
:
system msg for write_line failure : Bad file descriptor
distributed.nanny - WARNING - Restarting worker
[cli_2]: write_line error; fd=13 buf=:cmd=init pmi_version=1 pmi_subversion=1
##########################################

Any advise on why the worker won't stay running in the queue?  Thanks for taking the time to go through all of this!

**Comment by guillaumeeb** [comment]
Hi, several things here at first glance, distributed computing with Pbs and Dask can be difficult to grasp correctly :
- you're basically trying to redefined your own way of launching a Dask cluster on Pbs.
- Don't try to mix dask-mpi and dask-jobqueue, this is not advisable. If I read correctly your process, you're actually launching two different clusters, one with launch-scheduler, the other with launch-worker. Workers from second script will never connect to first cluster.
- dask-jobqueue is perfectly capable of launching batch scripts, without jupyter. If your algorithm needs a fixed worker pool however, prefer dask-mpi. 
- processes and cores kwargs are not used correctly. Processes should never be greater than cores. Moreover I recommand using local-directory with jobqueue too.
- It seems you've got an mpi library problem, which may explain why you loses workers. 

I recommend going through the docs of this project, and asking @mnlevy1981 or @jhamman some help to define correctly your need and scripts.

**Comment by vprzybylo** [comment]
Thanks for the help.  I have played around with a dask-jobqueue python script to submit my job since the workers stay online using qsub to start an interactive jupyter notebook (as I was doing before) but they now die after starting the main parallel job.  Here is the short script (edits made from your advice).  I thought the scheduler file should be updated once the cluster variable is declared but the time stamp on scheduler.json doesn't change and there is no dask-scheduler under name in the queue, only the workers. Not sure if this is right..

from dask_jobqueue import PBSCluster
from dask.distributed import Client
cluster = PBSCluster(processes=6,
                     cores=24,
                     memory="100GB",
		     local_directory='$TMPDIR',
                     project='UALB0017',
                     queue='economy',
                     resource_spec='select=1:ncpus=36:ompthreads=4:mem=109G',
                     walltime='02:00:00')
print(cluster.job_script())
client = Client(cluster)
c = cluster.start_workers(10)
#client = Client(scheduler_file='/glade/scratch/$USER/scheduler.json')


^^Gets called from a bash script that then waits for the workers to come online and then run the main script (section of it seen below).  The workers come online but die off after running the main script.

1. Since it is using Pool for parallelization, should I be using mpi? if so, do I use the scheduler and worker scripts previously attached?
2. how do I know how many processes to insert into the Pool?  In the notebook I used to print out client and insert the resulting processes (usually something around 180)

##############################################
def main():
    n_CPUs = multiprocessing.cpu_count()
    chunksize=1
    pool = Pool(processes=6) #<-- how to pass the processes from client to here?

    parallel_clus = partial(lab.parallelize_clusters, save_plots=save_plots,
                            minor=minor,nclusters=nclusters, ncrystals=ncrystals, numaspectratios=numaspectratios, 
speedy=speedy, rand_orient=rand_orient, ch_dist=ch_dist)


    results = pool.map(parallel_clus, phio, chunksize = chunksize)

###############################################
There is really no error in the worker logs, they just close:
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.4:35377
distributed.worker - INFO - Waiting to connect to:     tcp://10.148.0.4:35377
distributed.nanny - INFO - Closing Nanny at 'tcp://10.148.9.34:52142'
distributed.worker - INFO - Stopping worker at tcp://10.148.9.34:39979

I'd like to make sure I am approaching things correctly since I am new to running in parallel on Cheyenne --  @mnlevy1981 or @jhamman

**Comment by vprzybylo** [comment]
Just following up that the above message was seen -- thanks! @mnlevy1981 @jhamman @guillaumeeb

**Comment by guillaumeeb** [comment]
> there is no dask-scheduler under name in the queue, only the workers. Not sure if this is right..

That's normal, scheduler is started inside your notebook.

> Since it is using Pool for parallelization, should I be using mpi? if so, do I use the scheduler and worker scripts previously attached?
> pool = Pool(processes=6) 

Are you using multiprocessing python module? Are you not using client or Dask API like Delayed, Futures or Dataframes?

I'm not sure you've grasp how to use Dask yet. Could you provide your whole script? I recommend going through Dask documentation.

**Comment by mnlevy1981** [comment]
@vprzybylo -- my dask experience is limited to inside a notebook... I don't know enough to help troubleshoot a non-Jupyter script invoking jobqueue. That said, if you're only using a script because

> I have tried using qsub to start an interactive job for jupyter lab in prior attempts (see pangeo issue #352) but if the connection is dropped, the job disappears from the queue and I am unable to ssh back in to receive the output data.

Then maybe we could use a google hangout or something to try to figure out how to get the client scripts in [NCAR/pangeo](https://github.com/NCAR/pangeo) working. I didn't mean to give up on the issues you raised about it, it was just hard to figure out without being able to recreate the problem myself... so if you we can talk through the troubleshooting while you share your screen I bet we can get to the bottom of it. My email is mlevy [at] ucar.edu and I should be free anytime this week.

**Comment by mnlevy1981** [comment]
Let's chat tomorrow - I'm usually at my desk my 9:00 (Mountain), so maybe say 9:30 or later? Just send me an email and let me know what works best for you.

**Comment by guillaumeeb** [comment]
> I am using Pythons multiprocessing for parallelization

That sounds wrong to me, but maybe I'm mistaken. You should be using only Dask API exept if you have a very specific use case.

Could you show a simplified version of things your doing after the main() function definition?

**Comment by vprzybylo** [comment]
Hi @mnlevy1981, since you were able to help me earlier with the "launch-pangeo-on-cheyenne" script to start an interactive jupyter lab I have some followup questions.  I am using dask delayed as seen below to parallelize my code but I am still exceeding the wall clock max time limit of 12 hours on the economy or regular queue for Cheyenne.  I just submitted a job with 2 nodes and 36 cores but I am unfamiliar with the magnitude of these that I should be using for my job and am not sure exactly how long it should take to finish since this is my first attempt at parallelizing so I don't really have a gauge on runtime or resource use yet.

In ~/.config/dask/jobqueue.yaml I am using the default 36 cores for the workers which I am scaling to 15 (i.e., cluster.scale(15)) in the notebook.

Also, when I start the dask dashboard to look at the worker processes it is blank which concerns me that maybe the workers aren't connecting correctly but I can print out the futures that are completing in the notebook.  I am also writing them to a file to see how far it makes it if I loose connection to the notebook (since the print statements don't update upon ssh'ing back in) but they don't write to the file until after the job completes it looks like.  Same with the dask-worker.o files.  They are clearly running from the print statements I am outputting but don't update until the job exceeds the wall clock.

Please let me know if it is easier to chat on google hangouts again!

##########################################################
...main call...

    output = [delayed(lab.parallelize_clusters)(phio=x, save_plots=save_plots, 
                        minor=minor,nclusters=nclusters, ncrystals=ncrystals, numaspectratios=numaspectratios,
                        speedy=speedy, rand_orient=rand_orient, ch_dist=ch_dist) for x in phio]

    results = client.compute(output)
     
    count = 0
    with open('futures_complete.txt', 'w') as f3:
        for future, result in as_completed(results, with_results=True):
            print('completed: ', count)            
            f3.write(count)
            count += 1
    
    results = client.gather(results)

**Comment by guillaumeeb** [comment]
Hi @vprzybylo, 

If the original issue has been fixed with @mnlevy1981, it would be better to close it with some explanation.

Then I encourage you to open another issue if needed, or directly talk with @mnlevy1981.

Thanks.

**Comment by vprzybylo** [comment]
@guillaumeeb, yes sorry for posting here again.  Pulling the ./launch-pangeo-on-cheyenne script from the pangeo client-side github page reset something that had been altered to correctly set the hostname to ssh back in and start jupyter lab. 

And after upping my core/node count from the job submitted yesterday it did finish under the 12 hour wall clock time, so no need to start a new issue with @mnlevy1981.  

Additionally, the dashboard status task stream seems to be working ok after waiting awhile once submitting the job.  This includes checking the profile which now shows different colors for the different functions.  And for the task stream there is completely filled in solid color for the main function for every worker, yet the bytes stored and tasks processing graph is blank except for the initial spike upon queue submission.  Regardless, I have the output I need and should be able to proceed without opening up a new issue for a longer run time.

I appreciate the response!

**Comment by guillaumeeb** [comment]
Ok, glad you were able to deal with your problems. Closing this issue now. Do not hesitate if you have some problem with dask-jobqueue.
